sdks/python/apache_beam/runners/portability/flink_runner.py

```diff
@@ -0,0 +1,127 @@
+DEV_FLINK_JOB_SERVER_TEMPLATE = '{project_root}/runners/flink/{flink_version}/job-server/build/libs/beam-runners-flink-{flink_version}-job-server-{beam_version}-SNAPSHOT.jar'
+JAR_CACHE = os.path.expanduser("~/.apache_beam/cache")
+
+
+class FlinkRunner(portable_runner.PortableRunner):
+  def default_job_server(self, options):
+    return FlinkJarJobServer(options)
+
+
+class FlinkRunnerOptions(pipeline_options.PipelineOptions):
+  @classmethod
+  def _add_argparse_args(cls, parser):
+    parser.add_argument('--flink_master_url', default='[local]')
+    parser.add_argument('--flink_version',
+                        default=PUBLISHED_FLINK_VERSIONS[-1],
+                        choices=PUBLISHED_FLINK_VERSIONS,
+                        help='Flink version to use.')
+    parser.add_argument('--flink_job_server_jar',
+                        help='Path or URL to a flink jobserver jar.')
+
+
+
+class FlinkJarJobServer(job_server.SubprocessJobServer):
+  def __init__(self, options):
+    super(FlinkJarJobServer, self).__init__()
+    options = options.view_as(FlinkRunnerOptions)
+    self._jar = self.local_jar(self.flink_jar(options))
+    self._master_url = options.flink_master_url
+
+  def flink_jar(self, options):
+    if options.flink_job_server_jar:
+      return options.flink_job_server_jar
+    else:
+      if beam_version.endswith('.dev'):
+        # TODO: Attempt to use nightly snapshots?
+        project_root = os.path.sep.join(__file__.split(os.path.sep)[:-6])
+        dev_path = DEV_FLINK_JOB_SERVER_TEMPLATE.format(
+            project_root=project_root,
+            flink_version=options.flink_version,
+            beam_version=beam_version[:-4])
+        if os.path.exists(dev_path):
+          logging.warning(
+              'Using pre-built flink job server snapshot at %s', dev_path)
+          return dev_path
+        else:
+          raise RuntimeError(
+              'Please build the flink job server with '
+              'cd %s; ./gradlew runners:flink:%s:job-server:shadowJar' % (
+                  os.path.abspath(project_root), options.flink_version))
+      else:
+        return MAVEN_FLINK_JOB_SERVER_TEMPLATE.format(
+            maven_repo=MAVEN_REPOSITORY,
+            flink_version=options.flink_version,
+            beam_version=beam_version)
+
+  def local_jar(self, url):
+    if os.path.exists(url):
+      return url
+    else:
+      cached_jar = os.path.join(JAR_CACHE, os.path.basename(url))
+      if not os.path.exists(cached_jar):
+        if not os.path.exists(JAR_CACHE):
+          os.makedirs(JAR_CACHE)
+        try:
+          url_read = urlopen(url)
+          with open(cached_jar + '.tmp', 'wb') as jar_write:
+            shutil.copyfileobj(url_read, jar_write, length=1 << 20)
+          os.rename(cached_jar + '.tmp', cached_jar)
+        except URLError as e:
+          raise RuntimeError(
+              'Unable to fetch remote flink job server jar at %s: %s' % (url, e))
```

Comment:
Cleanup `.tmp` jar in case of errors during fetching? Retry routine in case of unreliable repository servers? May be overkill at this point.